source/blog/2015-05-08-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,216 @@
+>>
+```
+
+In this session, we extend the RSpec DSL on to a blank object and then call
+`o.expect(3).to Object.new`. This blows up because our new object does not implement
+the `matches?` method.
+
+In RSpec expectations, most matchers inherit from a class called `BaseMatcher`
+which causes `matches?` to delegate to a method called `match`. Some matchers have
+a very simple definition of `match`. For example, here's the definition of `match`
+for the `eq` matcher:
+
+```ruby
+def match(expected, actual)
+  actual == expected
+end
+```
+
+Some matchers, however, have a more complicated match protocol. This more complicated
+protocol is called the **composable** matcher protocol, which is what we'll discuss next.
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for some/all of it's
+arguments. The `match` matcher is perhaps an obvious place to start. In RSpec 2.xx the match
+matcher was really only used match regexes, and its match definition looked thus:
+
+```ruby
+def match(expected, actual)
+  actual.match expected
+end
+```
+
+Now obviously, one could extend the capabilites by providing other objects that provided a
+definition of match, but that's besides the point.
+
+The new definition of the match matcher's match method is thus:
+
+```ruby
+def match(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key important difference is that call to `values_match?` which is the core implementation
+of the new matching protocol in RSpec. The list of matchers which now reference the values_match?
+method are (as of RSpec expectations 3.3.0 development, not including mocks):
+
+* `change`
+* `contain_exactly`
+* `have_attributes`
+* `include`
+* `match`
+* `output`
+* `raise_error`
+* `start_or_end_with`
+* `throw_symbol`
+* `yield`
+
+Let's take a look at what `values_match?` does.
+
+`values_match?` is the RSpec match definition that allows matchers, and other objects, to compose.
+The implementation looks like this:
+
+```ruby
+def self.values_match?(expected, actual)
+  if Hash === actual
+    return hashes_match?(expected, actual) if Hash === expected
+  elsif Array === expected && Enumerable === actual && !(Struct === actual)
+    return arrays_match?(expected, actual.to_a)
+  end
+
+  return true if expected == actual
+
+  begin
+    expected === actual
+  rescue ArgumentError
+    # Some objects, like 0-arg lambdas on 1.9+, raise
+    # ArgumentError for `expected === actual`.
+    false
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is return `true` if expected === actual. An easy
```

Comment:
The next thing is `return true if expected == actual`, not `expected === actual`.

The `==` check is there because `/foo/ === /foo/` returns false as does `MyClass === MyClass` -- so we have to check normal equality as well and not rely solely on `===`.